File: src/cpp/llnms-viewer/utilities/NetworkUtilities.cpp
/**
 * @file    NetworkUtilities.cpp
 * @author  Marvin Smith
 * @date    1/11/2014
*/
#include "NetworkUtilities.hpp"

#include <boost/regex.hpp>

/**
 * Check if a network address is a valid ip4 address format
*/
bool isValidIP4AddressFormat( const std::string& ip4_address ){

    boost::regex expression("(\\d{1,3}(\\.\\d{1,3}){3})");
    return boost::regex_match( ip4_address, expression );  
}


/**
 * Check if string is in the proper format for a network hostname
*/
bool isValidHostnameFormat( std::string const& hostname ){

    boost::regex expression("[a-zA-Z]+([\\-\\.]*[a-zA-Z]+)*");
    return boost::regex_match( hostname, expression );

}

File: src/core/python/llnms/utility/Network_Utilities.py
__author__ = 'Marvin Smith'

#  LLNMS Libraries
from ..Globals import *

#  Python Libraries
import re

# ---------------------- #
# -      IP Type       - #
# ---------------------- #
class IP_Address_Type(object):

    #  Value
    UNKNOWN = -1
    IPV4    = 0
    IPV6    = 1

    # ----------------------------- #
    # -         To String         - #
    # ----------------------------- #
    def To_String(self, value ):

        #  If IPV4
        if value == self.IPV4:
            return 'ipv4'

        #  If IPV6
        if value == self.IPV6:
            return 'ipv6'

        #  Return
        return 'unknown'

    # ----------------------------- #
    # -       From String         - #
    # ----------------------------- #
    def From_String(self, value):

        # IP4
        if value == 'ipv4' or value == 'ip4' or value == 'ip4-address':
            return IP_Address_Type.IPV4

        #  IP6
        if value == 'ipv6' or value == 'ip6' or value == 'ip6-address':
            return IP_Address_Type.IPV6

        #  Unknown
        return IP_Address_Type.UNKNOWN

    # ------------------------------- #
    # -       Check if Valid        - #
    # ------------------------------- #
    def Is_Valid(self, value):

        if value == IP_Address_Type.IPV4:
            return True
        if value == IP_Address_Type.IPV6:
            return True
        return False

# ---------------------------------------------- #
# -     Make sure network input is valid       - #
# ---------------------------------------------- #
def Is_Valid_IP4_Address( address ):

    #  Check none
    if address is None:
        return False

    #  Regex Check
    prog = re.compile(LLNMS_IP4_REGEX_PATTERN)
    if prog.match(address) is None:
        return False

    #  Split on period
    comps = address.split('.')
    if len(comps) != 4:
        return False

    #  Check the range
    for x in xrange(0,4):

        #  Make sure the components are >= 0 and <= 255
        if int(comps[x]) < 0 or int(comps[x]) > 255:
            return False

    return True